src/analysis/normed_space/inner_product.lean

```diff
@@ -642,6 +660,66 @@ end
 
 end basic_properties
 
+section orthonormal_sets
+variables {ι : Type*} (𝕜)
+
+include 𝕜
+
+/-- An orthonormal set of vectors in an `inner_product_space` -/
+def orthonormal (v : ι → E) : Prop :=
+(∀ i, ∥v i∥ = 1) ∧ (∀ {i j}, i ≠ j → ⟪v i, v j⟫ = 0)
+
+omit 𝕜
+
+variables {𝕜}
+
+/-- `if ... then ... else` characterization of a set of vectors being orthonormal.  (Inner product
+equals Kronecker delta.) -/
+lemma orthonormal_iff_ite {v : ι → E} :
+  orthonormal 𝕜 v ↔ ∀ i j, ⟪v i, v j⟫ = if i = j then (1:𝕜) else (0:𝕜) :=
+begin
+  split,
+  { intros hv i j,
+    split_ifs,
+    { simp [h, inner_self_eq_norm_sq_to_K, hv.1] },
+    { exact hv.2 h } },
+  { intros h,
+    split,
+    { intros i,
+      have h' : ∥v i∥ ^ 2 = 1 ^ 2 := by simp [norm_sq_eq_inner, h i i],
+      have h₁ : 0 ≤ ∥v i∥ := norm_nonneg _,
+      have h₂ : (0:ℝ) ≤ 1 := by norm_num,
+      rwa eq_of_pow_two_eq_pow_two h₁ h₂ at h' },
+    { intros i j hij,
+      simpa [hij] using h i j } }
+end
+
+/-- The inner product of a linear combination of a set of orthonormal vectors with one of those
+vectors picks out the coefficient of that vector. -/
+lemma inner_finsupp_orthonormal {v : ι → E} (he : orthonormal 𝕜 v) (l : ι →₀ 𝕜) (i : ι) :
+  ⟪v i, finsupp.total ι E 𝕜 v l⟫ = l i :=
+by simp [finsupp.total_apply, finsupp.inner_sum, orthonormal_iff_ite.mp he]
+
```

Comment:
How about also adding the corresponding lemma with the linear combination on the left side?